runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
"lsof -p 4196 +E" enumerates all processes, so this operation can require dozen seconds on hosts with 20K processes. Any solutions which require to enumerate all tasks in procfs doesn't work, because a container should be frozen for this time.

> Because it's an unnamed pipe associated with a process outside of the
container (see my ‘lsof -p {pid} +E’ for how to find the other
processes associated with the pipe [1]).

If we will know the other processes associated with the pipe, how does this help us to restore the pipe? Our idea is that "Start" and "Restore" operations should be symmetrical. If we create descriptors before starting a container, we need to create the same set of descriptors before restoring the container.

>So you can see that its standard streams are wired up to runc's 3/4,
5/6, and 7/8, and that runC (as of v0.0.3) needs to do a better job
opening one-directional streams for this purpose ;).

Do you suggest to remember this mappings for each version of runC? And this approach doesn't work for cases when we set a few extra descriptors to Process.

Let's I show another example and you will try to explain how to handle it without externalFds.
```
conn, err := net.Dial("unix", "/monitor.sock")
sk, err := conn.File()

// runc start
process := libcontainer.Process{
    Args: ...,
    ExtraFiles: []*os.File{sk},
}
err = container.Start(&process)
```

```
$ ls -l /proc/2573/fd/385
lrwx------ 1 root root 64 Aug 28 12:01 /proc/2573/fd/385 -> socket:[3197260]
lrwx------ 1 root root 64 Aug 28 12:01 /proc/2573/fd/389 -> socket:[3197260]
lrwx------ 1 root root 64 Aug 28 12:02 /proc/2573/fd/405 -> socket:[3197298]
```

"externalFds" : [" socket:[3197260]"]

We create a socket, set it as a third descriptor for a process. Then  a process creates one more socket for it's own needs. /monitor.sock is unreachable from a container.

Then we decided to dump this container. CRIU will check that it know how to restore all sockets. It's possible to restore socket:[3197298], because it has been created inside CT, but  socket:[3197260] should be marked as external. Currently we get this information from externalFds.

Then we prepare a new socket and restore a container. All descriptors which were linked with  "socket:[3197260]" will be linked with the new socket.

```
// runc checkpoint
container, err := getContainer(context)
container.Checkpoint(options);
```

```
conn, err := net.Dial("udp", "monitor:8080")
sk, err := conn.File()

// runc restore
process := libcontainer.Process{
    Args: ...,
    ExtraFiles: []*os.File{sk},
}
err = container.Restore(&process, checkpointOpts)
```